File: 77.cpp
#include <algorithm>
#include <deque>
#include <iostream>
#include <map>
#include <set>
#include <unordered_map>
#include <unordered_set>
#include <vector>

using namespace std;
class Solution {
  vector<vector<int>> combs;

public:
  void bt(int n, int k, int idx, vector<int> &curr) {
    if (curr.size() == k) {
      combs.push_back(curr);
    } else {
      if (curr.size() + (n - idx) < k - 1) {
        return;
      }
      for (int i = idx; i <= n; ++i) {
        curr.push_back(i);
        bt(n, k, i + 1, curr);
        curr.pop_back();
      }
    }
  }
  vector<vector<int>> combine(int n, int k) {
    ios_base::sync_with_stdio(false);
    vector<int> curr;
    bt(n, k, 1, curr);
    return combs;
  }
};
int main() {}

File: 77.py
class Solution:
    def __init__(self):
        self.combs = []

    def gen_combs(self, n, k, idx, curr):
        if len(curr) == k:
            self.combs.append(curr)
            return
        else:
            if len(curr) + (n - idx) < k - 1:
                return
            for i in range(idx, n + 1):
                self.gen_combs(n, k, i + 1, curr + [i])
        return

    def combine(self, n: int, k: int) -> list[list[int]]:
        curr = []
        self.gen_combs(n, k, 1, curr)
        return self.combs


print(Solution().combine(4, 2))
